### PR TITLE
PWGHF: Implementing B0 → D- π+ → (π- K+ π-) π+ workflow

### PR DESCRIPTION
**Pull request by AlexBigO**
@fgrosa

### CODE REVIEW

**Comment by fgrosa** [comment]
> @fgrosa

Hi @AlexBigO, thanks a lot! It looks good to me, you just have to resolve the conflicts due to the Ds PR merged (https://github.com/AliceO2Group/O2Physics/pull/941)

**Comment by vkucera** [comment]
Hi @AlexBigO , there are still some problems with the formatting.

**Comment by AlexBigO** [comment]
Hi @fgrosa @vkucera , thanks for the comments. I first tried to solve the conflicts on github but it created some whitespace errors, so I solved them directly on my local branch and it looks like it passed the formatting checks.

File: PWGHF/TableProducer/HFB0ToDPiCandidateSelector.cxx [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+    }
+  }
+};
+
+WorkflowSpec defineDataProcessing(ConfigContext const& cfgc)
+{
+  WorkflowSpec workflow{};
+  workflow.push_back(adaptAnalysisTask<HfB0ToDPiCandidateSelector>(cfgc, TaskName{"hf-b0-todpi-candidate-selector"}));
+  return workflow;
+}
```
Comment (vkucera) [inline PWGHF/TableProducer/HFB0ToDPiCandidateSelector.cxx:167]: Do not use `TaskName`.

File: PWGHF/TableProducer/HFCandidateCreatorB0.cxx [hunk 2]
```diff
@@ -0,0 +1,381 @@
...
+
+      const auto& secondaryVertex = df3.getPCACandidate();
+      trackParVar0.propagateTo(secondaryVertex[0], magneticField);
+      trackParVar1.propagateTo(secondaryVertex[0], magneticField);
+      trackParVar2.propagateTo(secondaryVertex[0], magneticField);
+
+      // D∓ → π∓ K± π∓
+      array<float, 3> pvecpiK = {track0.px() + track1.px(), track0.py() + track1.py(), track0.pz() + track1.pz()};
+      array<float, 3> pvecD = {pvecpiK[0] + track2.px(), pvecpiK[1] + track2.py(), pvecpiK[2] + track2.pz()};
+      auto trackpiK = o2::dataformats::V0(df3.getPCACandidatePos(), pvecpiK, df3.calcPCACovMatrixFlat(),
```
Comment (vkucera) [inline PWGHF/TableProducer/HFCandidateCreatorB0.cxx:130]: Follow naming conventions.
```suggestion
      array<float, 3> pVecPiK = {track0.px() + track1.px(), track0.py() + track1.py(), track0.pz() + track1.pz()};
      array<float, 3> pVecD = {pvecpiK[0] + track2.px(), pvecpiK[1] + track2.py(), pvecpiK[2] + track2.pz()};
```

File: PWGHF/TableProducer/HFCandidateCreatorB0.cxx [hunk 3]
```diff
@@ -0,0 +1,381 @@
...
+      int index1D = track1.globalIndex();
+      int index2D = track2.globalIndex();
+      // int charge = track0.sign() + track1.sign() + track2.sign();
+
+      // loop on D-
+      // D- → π- K+ π-
+      // we don't have direct access to D sign so we use the sign of the daughters (the pion track0 here)
+      if (track0.sign() < 0) {
+        // loop over pions
+        for (auto& trackPion : tracks) {
```
Comment (vkucera) [inline PWGHF/TableProducer/HFCandidateCreatorB0.cxx:145]:
```suggestion
        for (auto const& trackPion : tracks) {
```

File: PWGHF/TableProducer/HFCandidateCreatorB0.cxx [hunk 4]
```diff
@@ -0,0 +1,381 @@
...
+            continue;
+          }
+          // we only keep pi+ to combine them with Dminus and reconstruct B0
+          if (trackPion.sign() < 0) {
+            continue;
+          }
+
+          hPtPion->Fill(trackPion.pt());
+          array<float, 3> pvecPion;
+          auto trackParVarPi = getTrackParCov(trackPion);
```
Comment (vkucera) [inline PWGHF/TableProducer/HFCandidateCreatorB0.cxx:161]:
```suggestion
          hPtPion->Fill(trackPion.pt());
          array<float, 3> pVecPion;
          auto trackParVarPi = getTrackParCov(trackPion);
```

File: PWGHF/TableProducer/HFCandidateCreatorB0.cxx [hunk 5]
```diff
@@ -0,0 +1,381 @@
...
+               aod::McParticles const& particlesMC)
+  {
+    int indexRec = -1;
+    int8_t sign = 0;
+    int8_t flag = 0;
+    int8_t origin = 0;
+    int8_t debug = 0;
+
+    // Match reconstructed candidates.
+    for (auto& candidate : candidates) {
```
Comment (vkucera) [inline PWGHF/TableProducer/HFCandidateCreatorB0.cxx:322]:
```suggestion
    for (auto const& candidate : candidates) {
```

**Review by vkucera — CHANGES_REQUESTED** [review]
Hi @AlexBigO , in general there many cases of `const` missing and you are using an obsolete task structure with `doMC` which does not work on AliHyperloop.

**Comment by vkucera** [comment]
> Hi @fgrosa @vkucera , thanks for the comments. I first tried to solve the conflicts on github but it created some whitespace errors, so I solved them directly on my local branch and it looks like it passed the formatting checks.

Looking at the latest version, the CMakeLists files are still a mess.

**Comment by vkucera** [comment]
@TimoWilken Why is the copyright check looking only at C++ files?

**Comment by vkucera** [comment]
Add your workflows in the validation framework.

**Comment by TimoWilken** [comment]
@vkucera, good question! Seems like CMake files generally have the copyright notice too, so I've added them in https://github.com/alisw/ali-bot/pull/1199.

**Comment by vkucera** [comment]
> @vkucera, good question! Seems like CMake files generally have the copyright notice too, so I've added them in [alisw/ali-bot#1199](https://github.com/alisw/ali-bot/pull/1199).

Thanks!

File: PWGHF/TableProducer/HFCandidateCreatorB0.cxx [hunk 6]
```diff
@@ -0,0 +1,381 @@
...
+                                   dCand.index2_as<aod::BigTracksMC>()};
+      // B0 → D- π+ → (π- K+ π-) π+
+      // Printf("Checking B0 → D- π+");
+      indexRec = RecoDecay::getMatchedMCRec(particlesMC, arrayDaughters, pdg::Code::kB0, array{-kPiPlus, +kKPlus, -kPiPlus, +kPiPlus}, true, &sign, 2);
+      if (indexRec > -1) {
+        // B0 → D- π+
+        // Printf("Checking B0 → D- π+");
+        indexRec = RecoDecay::getMatchedMCRec(particlesMC, arrayDaughtersD, -pdg::Code::kDPlus, array{-kPiPlus, +kKPlus, -kPiPlus}, true, &sign, 1);
+        if (indexRec > -1) {
+          flag = BIT(hf_cand_b0::DecayType::B0ToDPi);
```
[the remark this answers is not in the document]
Comment (AlexBigO) [reply PWGHF/TableProducer/HFCandidateCreatorB0.cxx:343]: What do you mean?

[the remark this answers is not in the document]
Comment (vkucera) [reply PWGHF/TableProducer/HFCandidateCreatorB0.cxx:343]: You don't multiply by `sign` here, whereas on line 363 you do.

[the remark this answers is not in the document]
Comment (AlexBigO) [reply PWGHF/TableProducer/HFCandidateCreatorB0.cxx:343]: Now I see what you mean. Indeed, thanks.

**Comment by AlexBigO** [comment]
Thanks for the comments @vkucera , I am gonna push a new version (that has been tested on a draft PR and that should pass the format tests and compile, hopefully).

File: PWGHF/Tasks/taskB0.cxx [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+  }
+
+  Filter filterSelectCandidates = (aod::hf_selcandidate_b0::isSelB0ToDPi >= selectionFlagB0);
+
+  void process(soa::Join<aod::Collisions, aod::CentRun2V0Ms>::iterator const& collision, soa::Filtered<soa::Join<aod::HfCandB0, aod::HFSelB0ToDPiCandidate>> const& candidates, soa::Join<aod::HfCandProng3, aod::HFSelDplusToPiKPiCandidate> const&, aod::BigTracks const&)
+  {
+    float centrality = collision.centRun2V0M();
+    registry.fill(HIST("hCentrality"), centrality);
+
+    for (auto& candidate : candidates) {
```
Comment (vkucera) [inline PWGHF/Tasks/taskB0.cxx:72]:
```suggestion
    for (auto const& candidate : candidates) {
```

File: PWGHF/Tasks/taskB0.cxx [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+    registry.add("hThetaStarRecBg", "B^{0} candidates (unmatched);B^{0} #cos(#theta^{*});entries", {HistType::kTH2F, {{110, -1.1, 1.1}, {(std::vector<double>)bins, "#it{p}_{T} (GeV/#it{c})"}}});
+  }
+
+  Filter filterSelectCandidates = (aod::hf_selcandidate_b0::isSelB0ToDPi >= selectionFlagB0);
+
+  void processMC(soa::Filtered<soa::Join<aod::HfCandB0, aod::HFSelB0ToDPiCandidate, aod::HfCandB0MCRec>> const& candidates,
+                 soa::Join<aod::McParticles, aod::HfCandB0MCGen> const& particlesMC, aod::BigTracksMC const& tracks, aod::HfCandProng3 const&)
+  {
+    // MC rec
+    for (auto& candidate : candidates) {
```
Comment (vkucera) [inline PWGHF/Tasks/taskB0.cxx:173]:
```suggestion
    for (auto const& candidate : candidates) {
```

File: PWGHF/Tasks/taskB0.cxx [hunk 9]
```diff
@@ -0,0 +1,270 @@
...
+        registry.fill(HIST("hCPADRecBg"), candD.cpa(), candidate.pt());
+        registry.fill(HIST("hDecLengthDRecBg"), candD.decayLength(), candidate.pt());
+        registry.fill(HIST("hChi2PCARecBg"), candidate.chi2PCA(), candidate.pt());
+        // registry.fill(HIST("hThetaStarRecBg"), candidate.cosThetaStar(), candidate.pt());
+      }
+    } // rec
+
+    // MC gen. level
+    // Printf("MC Particles: %d", particlesMC.size());
+    for (auto& particle : particlesMC) {
```
Comment (vkucera) [inline PWGHF/Tasks/taskB0.cxx:228]:
```suggestion
    for (auto const& particle : particlesMC) {
```

File: PWGHF/TableProducer/HFCandidateCreatorB0.cxx [hunk 10]
```diff
@@ -0,0 +1,372 @@
...
+                                  candD.index2_as<aod::BigTracksMC>(),
+                                  candidate.index1_as<aod::BigTracksMC>()};
+      auto arrayDaughtersD = array{candD.index0_as<aod::BigTracksMC>(),
+                                   candD.index1_as<aod::BigTracksMC>(),
+                                   candD.index2_as<aod::BigTracksMC>()};
+      // B0 → D- π+ → (π- K+ π-) π+
+      // Printf("Checking B0 → D- π+");
+      indexRec = RecoDecay::getMatchedMCRec(particlesMC, arrayDaughters, pdg::Code::kB0, array{-kPiPlus, +kKPlus, -kPiPlus, +kPiPlus}, true, &sign, 2);
+      if (indexRec > -1) {
+        // B0 → D- π+
```
Comment (vkucera) [inline PWGHF/TableProducer/HFCandidateCreatorB0.cxx:332]:
```suggestion
        // D- → π- K+ π-
```

**Review by vkucera — CHANGES_REQUESTED** [review]
There are a few remaining things to fix.
Make the PR in the validation framework.

**Review by vkucera — CHANGES_REQUESTED** [review]
There are many unused configurables in the selector. Do you have plans to add their implementation later? Especially `TPCNClsFindablePIDCut` seems obsolete and not used in other selectors either.

**Comment by AlexBigO** [comment]
Hey @vkucera. Indeed, the remaining configurables are for the PID that I did not implement yet, but I plan on adding it later. I will also remove the obsolete configurable, thanks.

**Comment by vkucera** [comment]
@TimoWilken Some checks are completely missing from the list. Do you know why that happened?

**Comment by vkucera** [comment]
> Hey @vkucera. Indeed, the remaining configurables are for the PID that I did not implement yet, but I plan on adding it later. I will also remove the obsolete configurable, thanks.

Thanks

**Review by vkucera — APPROVED** [review]

**Comment by vkucera** [comment]
@AlexBigO , it seems the CI checks have been having some troubles recently. Can you try to rebase and push to restart them?

**Review by vkucera — APPROVED** [review]

**Pull request merged**